File: Programmers/2/12978. 배달/배달.java
import java.util.*;

class Solution {
    static class Node {
        int to, cost;
        Node(int to, int cost) {
            this.to = to;
            this.cost = cost;
        }
    }

    public int solution(int N, int[][] road, int K) {
        List<List<Node>> graph = new ArrayList<>();
        
        for (int i = 0; i <= N; i++)
            graph.add(new ArrayList<>());

        for (int[] r : road) {
            int a = r[0], b = r[1], c = r[2];
            graph.get(a).add(new Node(b, c));
            graph.get(b).add(new Node(a, c));
        }

        int[] dist = new int[N + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[1] = 0;

        PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(n -> n.cost));
        pq.add(new Node(1, 0));

        while (!pq.isEmpty()) {
            Node curr = pq.poll();

            if (curr.cost > dist[curr.to]) continue;

            for (Node next : graph.get(curr.to)) {
                int newCost = curr.cost + next.cost;

                if (newCost < dist[next.to]) {
                    dist[next.to] = newCost;
                    pq.add(new Node(next.to, newCost));
                }
            }
        }

        int answer = 0;
        for (int i = 1; i <= N; i++) {
            if (dist[i] <= K) answer++;
        }

        return answer;
    }
}

File: Programmers/2/12978. 배달/배달.cpp
#include <iostream>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

struct Node
{
    int to, cost;
};

int solution(int N, vector<vector<int> > road, int K) {
    vector<vector<Node>> graph(N + 1);
    
    for (auto &r : road)
    {
        int a = r[0], b = r[1], c = r[2];
        
        graph[a].push_back({b, c});
        graph[b].push_back({a, c});
    }
    
    vector<int> dist(N + 1, INT_MAX);
    dist[1] = 0;
    
    priority_queue<pair<int,int>,
                  vector<pair<int,int>>,
                  greater<pair<int,int>>
    > pq;
    
    pq.push({0, 1});
    
    while (!pq.empty())
    {
        auto [cost, node] = pq.top();
        pq.pop();
        
        if (cost > dist[node]) continue;
        
        for (auto &next : graph[node])
        {
            int newCost = cost + next.cost;
            
            if (newCost < dist[next.to])
            {
                dist[next.to] = newCost;
                pq.push({newCost, next.to});
            }
        }
    }
    
    int answer = 0;
    for (int i = 1; i <= N; i++)
    {
        if (dist[i] <= K)
            answer++;
    }
    
    return answer;
}